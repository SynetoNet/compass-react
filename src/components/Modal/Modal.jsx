import React from "react";
import PropTypes from "prop-types";
// import classNames from "classnames";

import BSModal from "react-bootstrap/Modal";
import Button from "../Button/Button";
import "./Modal.scss";

class Modal extends React.Component {
  static Header = BSModal.Header;
  static Title = BSModal.Title;
  static Body = ({ children, ...props }) => {
    return (
      <>
        <div className="modal-shadow-top" />
        <BSModal.Body {...props}>
          <div className="modal-body-inner">{children}</div>
        </BSModal.Body>
        <div className="modal-shadow-bottom" />
      </>
    );
  };
  static Footer = BSModal.Footer;

  render() {
    const {
      children,
      title,
      content,
      actions,
      closeButton,
      ...props
    } = this.props;
    // const classes = classNames({});

    if (children) {
      return (
        <BSModal {...props} onEntering={this.onEnter}>
          {children}
        </BSModal>
      );
    }

    return (
      <BSModal {...props} onEntering={this.onEnter}>
        {title && (
          <Modal.Header closeButton={closeButton}>
            <Modal.Title>{title}</Modal.Title>
          </Modal.Header>
        )}
        {content && <Modal.Body>{content}</Modal.Body>}
        <Modal.Footer>{this._renderFooter(actions)}</Modal.Footer>
      </BSModal>
    );
  }

  _renderFooter = ({ primary, secondary, tertiary }) => {
    const elements = [];

    if (tertiary) {
      elements.push(
        <div className="modal-footer_tertiary" key="tertiary">
          <Button
            role="tertiary"
            onClick={tertiary.onClick}
            className="px-0"
            disabled={tertiary.disabled ? true : false}
          >
            {tertiary.label}
          </Button>
        </div>
      );
    }

    if (secondary) {
      elements.push(
        <Button
          role="secondary"
          onClick={secondary.onClick || this.props.onHide}
          key="secondary"
          disabled={secondary.disabled ? true : false}
        >
          {secondary.label}
        </Button>
      );
    }

    if (primary) {
      elements.push(
        <Button
          onClick={primary.onClick}
          key="primary"
          className="ml-2"
          disabled={primary.disabled ? true : false}
        >
          {primary.label}
        </Button>
      );
    }

    return elements;
  };

  onEnter = modal => {
    const { onEntering } = this.props;

    // support stacked modals
    const START_Z_INDEX = 1040;
    const nrModals = document.querySelectorAll(".modal.show").length;
    modal.style.zIndex = START_Z_INDEX + 2 * nrModals + 1;
    const modalBackdrop = document.querySelectorAll(".modal-backdrop.show")[
      nrModals - 1
    ];
    modalBackdrop.style.zIndex = START_Z_INDEX + 2 * nrModals;

    onEntering && onEntering();
  };
}

const actionsShape = {
  label: PropTypes.string,
  onClick: PropTypes.func,
  disabled: PropTypes.bool
};
Modal.propTypes = {
  /**
   * Footer actions buttons: tertiary (left aligned), secondary & primary (right-aligned) in the form of:
   *
   * `{ label: "button text", onClick: () => {} }`
   *
   * Defaults to secondary button with **`Close`** label
   */
  actions: PropTypes.shape({
    primary: PropTypes.shape(actionsShape),
    secondary: PropTypes.shape(actionsShape),
    tertiary: PropTypes.shape(actionsShape)
  }),
  /** String or any component to be rendered */
  content: PropTypes.node,
  /** String or any component to be rendered */
  title: PropTypes.node,
  backdrop: PropTypes.oneOf([false, true, "static"]),
  size: PropTypes.oneOf(["sm", "lg"])
};

Modal.defaultProps = {
  actions: {
    secondary: {
      label: "Close"
    }
  },
  backdrop: "static"
};

export default Modal;
