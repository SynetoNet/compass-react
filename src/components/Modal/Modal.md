#### **Simple Modal**

```jsx
import { useState } from "react";
import { Button } from "../../index";

const SimpleModal = () => {
  const [showModal, toggleModal] = useState(false);

  return (
    <>
      <Button onClick={() => toggleModal(true)}>Toggle modal</Button>
      <Modal
        show={showModal}
        onHide={() => toggleModal(false)}
        content="Modal content"
        title="Modal title"
      />
    </>
  );
};

<SimpleModal />;
```

#### **Action buttons**

```jsx
import { useState } from "react";
import { Button } from "../../index";

const ActionsModal = () => {
  const [showModal, toggleModal] = useState(false);

  return (
    <>
      <Button onClick={() => toggleModal(true)}>Toggle modal</Button>
      <Modal
        show={showModal}
        onHide={() => toggleModal(false)}
        content="Modal content"
        title="Modal title"
        actions={{
          primary: {
            label: "Submit",
            onClick: () => console.log("Submit clicked")
          },
          secondary: {
            label: "Cancel"
          },
          tertiary: {
            label: "See documentation",
            onClick: () => console.log("See documentation clicked")
          }
        }}
      />
    </>
  );
};

<ActionsModal />;
```

#### **Scrollable Modal**

```jsx
import { useState } from "react";
import { Button } from "../../index";

const ScrollableModal = () => {
  const [showModal, toggleModal] = useState(false);

  return (
    <>
      <Button onClick={() => toggleModal(true)}>Toggle modal</Button>
      <Modal
        show={showModal}
        onHide={() => toggleModal(false)}
        scrollable
        content={
          <div>
            <div style={{ height: "100vh" }} />
            content
            <div style={{ height: "100vh" }} />
          </div>
        }
        title="Modal title"
      />
    </>
  );
};

<ScrollableModal />;
```

#### **Sizes**

```jsx
import { useState } from "react";
import { Button } from "../../index";

const SizedModal = () => {
  const [showModalSm, toggleModalSm] = useState(false);
  const [showModalMd, toggleModalMd] = useState(false);
  const [showModalLg, toggleModalLg] = useState(false);

  return (
    <>
      <Button onClick={() => toggleModalSm(true)}>Small modal</Button>
      <Modal
        show={showModalSm}
        size="sm"
        onHide={() => toggleModalSm(false)}
        scrollable
        content="..."
        title="Small Modal"
      />

      <i className="px-2" />

      <Button onClick={() => toggleModalMd(true)}>Normal modal</Button>
      <Modal
        show={showModalMd}
        onHide={() => toggleModalMd(false)}
        scrollable
        content="..."
        title="Normal Modal"
      />

      <i className="px-2" />

      <Button onClick={() => toggleModalLg(true)}>Large modal</Button>
      <Modal
        show={showModalLg}
        size="lg"
        onHide={() => toggleModalLg(false)}
        scrollable
        content="..."
        title="Large Modal"
      />
    </>
  );
};

<SizedModal />;
```

#### **Stacked Modals**

```jsx
import { useState } from "react";
import { Button } from "../../index";

const SmallModal = () => {
  const [show, toggle] = useState(false);
  return (
    <>
      <Button onClick={() => toggle(true)}>Third Modal</Button>
      <Modal show={show} size="sm" onHide={() => toggle(false)} content="..." />
    </>
  );
};

const MediumModal = () => {
  const [show, toggle] = useState(false);
  return (
    <>
      <Button onClick={() => toggle(true)}>Second modal</Button>
      <Modal
        show={show}
        onHide={() => toggle(false)}
        content={<SmallModal />}
      />
    </>
  );
};

const FirstModal = () => {
  const [show, toggle] = useState(false);
  return (
    <>
      <Button onClick={() => toggle(true)}>First modal</Button>
      <Modal
        show={show}
        size="lg"
        onHide={() => toggle(false)}
        content={<MediumModal />}
      />
    </>
  );
};

<FirstModal />;
```

#### **Extended usage**

```jsx
import { useState } from "react";
import { Button, Badge } from "../../index";

const ExtendedModal = () => {
  const [showModal, toggleModal] = useState(false);

  return (
    <>
      <Button onClick={() => toggleModal(true)}>Toggle modal</Button>
      <Modal show={showModal} onHide={() => toggleModal(false)}>
        <Modal.Header>
          <div
            style={{
              flex: 1,
              display: "flex",
              justifyContent: "space-between",
              alignItems: "center"
            }}
          >
            <Modal.Title>Custom title</Modal.Title>
            <Badge>Info Badge</Badge>
          </div>
        </Modal.Header>

        <Modal.Body>
          <p>Lorem ipsum dolor sit amet.</p>
        </Modal.Body>

        <Modal.Footer>
          <div
            style={{
              flex: 1,
              display: "flex",
              justifyContent: "space-between"
            }}
          >
            <Button color="danger" onClick={() => toggleModal(false)}>
              No way!!!
            </Button>
            <Button color="success" onClick={() => toggleModal(false)}>
              Sure, anytime
            </Button>
          </div>
        </Modal.Footer>
      </Modal>
    </>
  );
};

<ExtendedModal />;
```
