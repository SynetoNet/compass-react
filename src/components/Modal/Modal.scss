.modal-header,
.modal-footer {
  border: 0;
  padding: 1.5rem;
}

.modal-body {
  padding: 1rem 1.5rem;
}

.modal-footer_tertiary {
  display: flex;
  flex: 1;
}
