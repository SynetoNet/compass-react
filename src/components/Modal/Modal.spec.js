import React from "react";
import Modal from "./Modal";

import { configure, mount, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("Default Modal", () => {
  const example = <Modal show={true} backdrop={"static"} />;

  test("modal without header or content", () => {
    const component = mount(example);
    expect(component.exists(".modal-header")).toBeFalsy();
    expect(component.exists(".modal-body")).toBeFalsy();
    expect(component.exists(".btn-outline-primary")).toBeTruthy();
    expect(component.find(".btn-outline-primary").text()).toEqual("Close");
  });

  test("modal without header or content snapshop", () => {
    const component = shallow(example);
    expect(component).toMatchSnapshot();
  });
});

describe("Simple Modal", () => {
  const example = (
    <Modal show={true} title="modal title" content="modal content" />
  );

  test("modal with title and content", () => {
    const component = mount(example);
    expect(component.exists(".modal-header")).toBeTruthy();
    expect(component.exists(".modal-body")).toBeTruthy();
  });

  test("modal with title and content snapshop", () => {
    const component = shallow(example);
    expect(component).toMatchSnapshot();
  });
});

describe("Actions Modal", () => {
  const example = (
    <Modal
      show={true}
      actions={{
        primary: {
          label: "Primary"
        },
        secondary: {
          label: "Cancel"
        },
        tertiary: {
          label: "Tertiary"
        }
      }}
    />
  );

  test("modal with actions", () => {
    const component = mount(example);
    expect(component.exists(".btn-outline-primary")).toBeTruthy();
    expect(component.find(".btn-outline-primary").text()).toEqual("Cancel");
    expect(component.exists(".btn-primary")).toBeTruthy();
    expect(component.find(".btn-primary").last().text()).toEqual("Primary");
    expect(component.exists(".btn-link")).toBeTruthy();
    expect(component.find(".btn-link").text()).toEqual("Tertiary");
  });

  test("modal with actions snapshop", () => {
    const component = shallow(example);
    expect(component).toMatchSnapshot();
  });
});

describe("Extended Modal", () => {
  const example = (
    <Modal show={true}>
      <Modal.Header>
        <div className="custom-title">title</div>
      </Modal.Header>
      <Modal.Body>
        <p className="custom-body">body</p>
      </Modal.Body>
      <Modal.Footer>
        <div className="custom-footer">footer</div>
      </Modal.Footer>
    </Modal>
  );

  test("extended modal", () => {
    const component = mount(example);
    expect(component.exists(".custom-title")).toBeTruthy();
    expect(component.exists(".custom-body")).toBeTruthy();
    expect(component.exists(".custom-footer")).toBeTruthy();
  });

  test("extended modal snapshop", () => {
    const component = shallow(example);
    expect(component).toMatchSnapshot();
  });
});
