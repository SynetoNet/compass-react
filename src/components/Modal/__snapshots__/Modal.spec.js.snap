// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions Modal modal with actions snapshop 1`] = `
<Bootstrap(Modal)
  show={true}
>
  <ModalFooter>
    <div
      className="modal-footer_tertiary"
      key="tertiary"
    >
      <Button
        className="px-0"
        color="default"
        disabled={false}
        variant="tertiary"
        width="auto"
      >
        Tertiary
      </Button>
    </div>
    <Button
      color="default"
      disabled={false}
      key="secondary"
      variant="secondary"
      width="auto"
    >
      Cancel
    </Button>
    <Button
      className="ml-2"
      color="default"
      disabled={false}
      key="primary"
      variant="primary"
      width="auto"
    >
      Primary
    </Button>
  </ModalFooter>
</Bootstrap(Modal)>
`;

exports[`Default Modal modal without header or content snapshop 1`] = `
<Bootstrap(Modal)
  show={true}
>
  <ModalFooter>
    <Button
      color="default"
      disabled={false}
      key="secondary"
      variant="secondary"
      width="auto"
    >
      Close
    </Button>
  </ModalFooter>
</Bootstrap(Modal)>
`;

exports[`Extended Modal extended modal snapshop 1`] = `
<Bootstrap(Modal)
  show={true}
>
  <ModalHeader
    closeButton={false}
    closeLabel="Close"
  >
    <div
      className="custom-title"
    >
      title
    </div>
  </ModalHeader>
  <ModalBody>
    <p
      className="custom-body"
    >
      body
    </p>
  </ModalBody>
  <ModalFooter>
    <div
      className="custom-footer"
    >
      footer
    </div>
  </ModalFooter>
</Bootstrap(Modal)>
`;

exports[`Simple Modal modal with title and content snapshop 1`] = `
<Bootstrap(Modal)
  show={true}
>
  <ModalHeader
    closeButton={false}
    closeLabel="Close"
  >
    <ModalTitle>
      modal title
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    modal content
  </ModalBody>
  <ModalFooter>
    <Button
      color="default"
      disabled={false}
      key="secondary"
      variant="secondary"
      width="auto"
    >
      Close
    </Button>
  </ModalFooter>
</Bootstrap(Modal)>
`;
